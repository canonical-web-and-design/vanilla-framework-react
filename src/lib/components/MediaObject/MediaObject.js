import React from 'react';
import PropTypes from 'prop-types';

const MediaObject = (props) => {
  let image;
  let title;
  let description;
  let metadata;

  if (props.img) {
    image = (
      <img
        className={`p-media-object__image${props.round ? ' is-round' : ''}`}
        src={props.img.src}
        alt={props.img.alt}
      />
    );
  }

  if (props.title) {
    if (props.title.link) {
      title = (
        <h3 className="p-media-object__title">
          <a href={props.title.link}>{props.title.name}</a>
        </h3>
      );
    } else {
      title = (
        <h3 className="p-media-object__title">
          {props.title.name}
        </h3>
      );
    }
  }

  if (props.description) {
    description = (
      <p className="p-media-object__content">
        {props.description}
      </p>
    );
  }

  if (props.metadata) {
    const metadataItems = Object.keys(props.metadata).map((key) => {
      if (key === 'topic') {
        return (
          <li key={key} className="p-media-object__meta-list-item">
            {props.metadata[key]}
          </li>
        );
      }
      return (
        <li key={key} className={`p-media-object__meta-list-item--${key}`}>
          {props.metadata[key]}
        </li>
      );
    });

    metadata = (
      <ul className="p-media-object__meta-list">
        {metadataItems}
      </ul>
    );
  }

  return (
    <div className={props.large ? 'p-media-object--large' : 'p-media-object'}>
      {image}
      <div className="p-media-object__details">
        {title}
        {description}
        {metadata}
      </div>
    </div>
  );
};

MediaObject.defaultProps = {
  round: false,
  large: false,
  img: null,
  title: null,
  description: null,
  metadata: null,
};

MediaObject.propTypes = {
  round: PropTypes.bool,
  large: PropTypes.bool,
  img: PropTypes.shape({
    src: PropTypes.string,
    alt: PropTypes.string,
  }),
  title: PropTypes.shape({
    name: PropTypes.string,
    link: PropTypes.string,
  }),
  description: PropTypes.string,
  metadata: PropTypes.shape({
    topic: PropTypes.string,
    date: PropTypes.string,
    venue: PropTypes.string,
    location: PropTypes.string,
  }),
};

MediaObject.displayName = 'MediaObject';

export default MediaObject;
