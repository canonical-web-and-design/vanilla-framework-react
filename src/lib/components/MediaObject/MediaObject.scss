@import 'vanilla-framework/scss/base';
@import 'vanilla-framework/scss/patterns_media-object';
@import 'vanilla-framework/scss/utilities_off-screen';
@include vf-b-typography;
@include vf-b-media;
@include vf-p-media-object;
