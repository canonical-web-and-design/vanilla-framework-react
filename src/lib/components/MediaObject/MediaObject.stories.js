import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import MediaObject from './MediaObject';

storiesOf('Media Object', module)
  .add('Default',
    withInfo('The MediaObject component should be used to display events or articles.')(() => (
      <MediaObject
        img={{ src: 'http://placehold.it/72x72', alt: '' }}
        title={{ name: 'Event Title', link: '#' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua."
        metadata={{
          topic: 'CSS frameworks',
          date: '21 - 23 February 2017',
          venue: 'Venue 360',
          location: 'Santa Clara, CA',
        }}
      />),
    ),
  )

  .add('Round',
    withInfo('You can add the "round" prop to the MediaObject component to create a circular image style, which we recommend to be used for head shots of people.')(() => (
      <MediaObject
        round
        img={{ src: 'http://placehold.it/72x72', alt: '' }}
        title={{ name: 'Person Name', link: '#' }}
        description="Lorem ipsum dolor sit amet"
        metadata={{ location: 'London, UK' }}
      />),
    ),
  )

  .add('Large',
    withInfo('Add the "large" prop to display details of a single object on a page.')(() => (
      <MediaObject
        large
        img={{ src: 'http://placehold.it/96x96', alt: '' }}
        title={{ name: 'Title' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor
        incididunt ut labore et dolore magna aliqua."
      />),
    ),
  )

  .add('Examples',
    withInfo('All of the props are optional, allowing many levels of customisation.')(() => (
      <div>
        <div style={{ margin: '10px 0', borderBottom: 'solid 1px #ccc' }} />
        <MediaObject
          large
          round
          img={{ src: 'http://placehold.it/96x96', alt: '' }}
          metadata={{
            topic: 'CSS frameworks',
            date: '21 - 23 February 2017',
            location: 'Santa Clara, CA',
          }}
        />
        <div style={{ margin: '10px 0', borderBottom: 'solid 1px #ccc' }} />
        <MediaObject
          title={{ name: 'Event Title' }}
          description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
          tempor incididunt ut labore et dolore magna aliqua."
          metadata={{
            date: '21 - 23 February 2017',
            venue: 'Venue 360',
          }}
        />
        <div style={{ margin: '10px 0', borderBottom: 'solid 1px #ccc' }} />
        <MediaObject
          img={{ src: 'http://placehold.it/72x72', alt: '' }}
          title={{ name: 'Event Title', link: '#' }}
          description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
          tempor incididunt ut labore et dolore magna aliqua."
        />
        <div style={{ margin: '10px 0', borderBottom: 'solid 1px #ccc' }} />
        <MediaObject />
      </div>),
    ),
  );
