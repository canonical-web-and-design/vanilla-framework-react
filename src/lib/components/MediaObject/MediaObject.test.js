import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import MediaObject from './MediaObject';

describe('MediaObject component', () => {
  it('should not render anything if props object is empty', () => {
    const mediaObject = ReactTestRenderer.create(
      <MediaObject />,
    );
    const json = mediaObject.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render default style correctly', () => {
    const mediaObject = ReactTestRenderer.create(
      <MediaObject
        title={{ name: 'Event Title' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua."
      />,
    );
    const json = mediaObject.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render large prop correctly', () => {
    const mediaObject = ReactTestRenderer.create(
      <MediaObject
        large
        title={{ name: 'Event Title' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua."
      />,
    );
    const json = mediaObject.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render round prop correctly', () => {
    const mediaObject = ReactTestRenderer.create(
      <MediaObject
        round
        title={{ name: 'Event Title' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua."
      />,
    );
    const json = mediaObject.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render image correctly', () => {
    const mediaObject = ReactTestRenderer.create(
      <MediaObject
        img={{ src: 'http://placehold.it/72x72', alt: '' }}
        title={{ name: 'Event Title', link: '#' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua."
      />,
    );
    const json = mediaObject.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render title link correctly', () => {
    const mediaObject = ReactTestRenderer.create(
      <MediaObject
        title={{ name: 'Event Title', link: '#' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua."
      />,
    );
    const json = mediaObject.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render metadata correctly', () => {
    const mediaObject = ReactTestRenderer.create(
      <MediaObject
        title={{ name: 'Event Title', link: '#' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua."
        metadata={{
          topic: 'CSS frameworks',
          date: '21 - 23 February 2017',
          venue: 'Venue 360',
          location: 'Santa Clara, CA',
        }}
      />,
    );
    const json = mediaObject.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a MediaObject with several optional props correctly', () => {
    const mediaObject = ReactTestRenderer.create(
      <MediaObject
        round
        large
        title={{ name: 'Event Title', link: '#' }}
        description="Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua."
        metadata={{
          topic: 'CSS frameworks',
          date: '21 - 23 February 2017',
          venue: 'Venue 360',
          location: 'Santa Clara, CA',
        }}
      />,
    );
    const json = mediaObject.toJSON();
    expect(json).toMatchSnapshot();
  });
});
