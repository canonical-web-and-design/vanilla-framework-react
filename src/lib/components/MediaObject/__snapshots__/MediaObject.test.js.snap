// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaObject component should not render anything if props object is empty 1`] = `
<div
  className="p-media-object"
>
  <div
    className="p-media-object__details"
  />
</div>
`;

exports[`MediaObject component should render a MediaObject with several optional props correctly 1`] = `
<div
  className="p-media-object--large"
>
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      <a
        href="#"
      >
        Event Title
      </a>
    </h3>
    <p
      className="p-media-object__content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </p>
    <ul
      className="p-media-object__meta-list"
    >
      <li
        className="p-media-object__meta-list-item"
      >
        CSS frameworks
      </li>
      <li
        className="p-media-object__meta-list-item--date"
      >
        21 - 23 February 2017
      </li>
      <li
        className="p-media-object__meta-list-item--venue"
      >
        Venue 360
      </li>
      <li
        className="p-media-object__meta-list-item--location"
      >
        Santa Clara, CA
      </li>
    </ul>
  </div>
</div>
`;

exports[`MediaObject component should render default style correctly 1`] = `
<div
  className="p-media-object"
>
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      Event Title
    </h3>
    <p
      className="p-media-object__content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </p>
  </div>
</div>
`;

exports[`MediaObject component should render image correctly 1`] = `
<div
  className="p-media-object"
>
  <img
    alt=""
    className="p-media-object__image"
    src="http://placehold.it/72x72"
  />
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      <a
        href="#"
      >
        Event Title
      </a>
    </h3>
    <p
      className="p-media-object__content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </p>
  </div>
</div>
`;

exports[`MediaObject component should render large prop correctly 1`] = `
<div
  className="p-media-object--large"
>
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      Event Title
    </h3>
    <p
      className="p-media-object__content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </p>
  </div>
</div>
`;

exports[`MediaObject component should render metadata correctly 1`] = `
<div
  className="p-media-object"
>
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      <a
        href="#"
      >
        Event Title
      </a>
    </h3>
    <p
      className="p-media-object__content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </p>
    <ul
      className="p-media-object__meta-list"
    >
      <li
        className="p-media-object__meta-list-item"
      >
        CSS frameworks
      </li>
      <li
        className="p-media-object__meta-list-item--date"
      >
        21 - 23 February 2017
      </li>
      <li
        className="p-media-object__meta-list-item--venue"
      >
        Venue 360
      </li>
      <li
        className="p-media-object__meta-list-item--location"
      >
        Santa Clara, CA
      </li>
    </ul>
  </div>
</div>
`;

exports[`MediaObject component should render round prop correctly 1`] = `
<div
  className="p-media-object"
>
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      Event Title
    </h3>
    <p
      className="p-media-object__content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </p>
  </div>
</div>
`;

exports[`MediaObject component should render title link correctly 1`] = `
<div
  className="p-media-object"
>
  <div
    className="p-media-object__details"
  >
    <h3
      className="p-media-object__title"
    >
      <a
        href="#"
      >
        Event Title
      </a>
    </h3>
    <p
      className="p-media-object__content"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </p>
  </div>
</div>
`;
